File: pkg-py/src/biobouncer/_fuzzy.py
"""Fuzzy "did you mean" suggestions by bounded edit distance.

A pure, deterministic Levenshtein used only in cache and existence modes, where a
local set of valid ids is available. It is specified exactly so R and Python make
the identical suggestion: bounded edit distance with unit costs, ties broken by
the code-point-smallest candidate.
"""

from __future__ import annotations


def _bounded_levenshtein(a: str, b: str, k: int) -> int | None:
    """The Levenshtein distance between ``a`` and ``b`` if it is at most ``k``.

    Returns ``None`` when the distance exceeds ``k``. Unit insertion, deletion,
    and substitution costs, no transposition, case sensitive. A running row
    minimum stops the search early once every path already exceeds ``k``.
    """
    la, lb = len(a), len(b)
    if abs(la - lb) > k:
        return None
    previous = list(range(lb + 1))
    for i in range(1, la + 1):
        current = [i] + [0] * lb
        row_min = i
        ai = a[i - 1]
        for j in range(1, lb + 1):
            cost = 0 if ai == b[j - 1] else 1
            current[j] = min(
                previous[j] + 1, current[j - 1] + 1, previous[j - 1] + cost
            )
            if current[j] < row_min:
                row_min = current[j]
        if row_min > k:
            return None
        previous = current
    distance = previous[lb]
    return distance if distance <= k else None


def fuzzy_index(ids) -> dict[int, list[str]]:
    """Bucket candidate ids by length so the search can prune by length."""
    buckets: dict[int, list[str]] = {}
    for candidate in ids:
        buckets.setdefault(len(candidate), []).append(candidate)
    return buckets


def fuzzy_suggest(s: str, index: dict[int, list[str]], k: int) -> str | None:
    """The nearest candidate within edit distance ``k``, or ``None``.

    Only candidates whose length is within ``k`` of ``s`` can be within distance
    ``k``, so the search visits just those length buckets. Among the candidates at
    the smallest distance, the code-point-smallest is chosen, so R and Python
    always agree.
    """
    best: str | None = None
    best_distance = k + 1
    length = len(s)
    for candidate_length in range(length - k, length + k + 1):
        for candidate in index.get(candidate_length, ()):
            distance = _bounded_levenshtein(s, candidate, k)
            if distance is None:
                continue
            if distance < best_distance or (
                distance == best_distance and candidate < best
            ):
                best_distance = distance
                best = candidate
    return best

File: pkg-r/R/fuzzy.R
# Fuzzy "did you mean" suggestions by bounded edit distance. Used only in cache
# and existence modes, where a local set of valid ids is available. Specified to
# match _fuzzy.py exactly: Levenshtein with unit costs, the nearest candidate
# within the configured distance, ties broken by the code-point-smallest.

# The nearest id to s within the source's fuzzy edit distance, or NA. Only a
# candidate whose length is within the distance can match, so the search is
# limited to those. adist() computes the same unit-cost Levenshtein as the Python
# side, and radix order breaks ties by code point.
#
# When the source sets suggest.case_insensitive, distance is measured between
# lowercase forms so a difference in case costs no edits, matching the
# ignore_case branch of fuzzy_suggest(). The candidates keep their original
# spelling throughout, so the returned suggestion and the code-point tie-break
# are both on the snapshot's own spelling, exactly as in Python.
.fuzzy_suggest <- function(source, s, ids) {
  cfg <- source[["suggest"]]$fuzzy
  if (is.null(cfg)) {
    return(NA_character_)
  }
  k <- suppressWarnings(as.integer(cfg$max_distance))
  if (is.na(k) || length(ids) == 0L) {
    return(NA_character_)
  }
  cand <- ids[abs(nchar(ids) - nchar(s)) <= k]
  if (!length(cand)) {
    return(NA_character_)
  }
  ignore_case <- isTRUE(source[["suggest"]]$case_insensitive)
  probe <- if (ignore_case) tolower(s) else s
  target <- if (ignore_case) tolower(cand) else cand
  d <- as.integer(adist(probe, target))
  keep <- !is.na(d) & d <= k
  if (!any(keep)) {
    return(NA_character_)
  }
  cand <- cand[keep]
  d <- d[keep]
  best <- cand[d == min(d)]
  sort(best, method = "radix")[1]
}
